Migrating python documentation page to java

manishbansal8843 opened the pull request:
[This ](https://docs.couchbase.com/java-sdk/2.7/core-operations.html) page should be migrated from python to java

manishbansal8843 committed:
Migrating python documentation page to java

manishbansal8843 commented:
This is WIP. Kindly bear with me. #71

manishbansal8843 committed:
Added more changes

manishbansal8843 committed:
Migrated data structure section

manishbansal8843 committed:
Migrated raw byte concatenation section

manishbansal8843 commented:
@RichardSmedley I have done my best to migrate the document. Kindly review the PR..

manishbansal8843 committed:
Updated primitive operations section

programmatix commented on modules/ROOT/pages/core-operations.adoc:
```diff
@@ -210,19 +210,16 @@ The same is the case for `decr()`.
-51L
+bucket.counter("counter_id", 20, 100).content(); //100L
+bucket.counter("counter_id", 1).content(); //101L
+bucket.counter("counter_id", -51).content(); //51L
```
I'd expect 101-51 to be //50L though, not // 51L.

manishbansal8843 replied:
sorry. my mistake. I was supposed to do `bucket.counter("counter_id", -50).content(); //51L`. Will correct this.

programmatix commented on modules/ROOT/pages/core-operations.adoc:
```diff
@@ -239,12 +236,14 @@ xref:concurrent-mutations-cluster.adoc[CAS] values are not used with counter ope
-  cb.upsert('counter_id', value + increment_amount, cas=cas)
+JsonDocument counter = bucket.get("counter_id");
+long cas = document.cas();
+long value = document.content();
```
Can you please check these examples compile.  None of this looks correct: upsert doesn't take an execute, you get document with variable 'counter' but use 'document' instead, 'JsonDocument#content()' returns a JsonObject not a long.  You probably want to start with `JsonLongDocument counter = bucket.counter("counter_id");`

manishbansal8843 replied:
Yes. Will correct this.

programmatix commented on modules/ROOT/pages/core-operations.adoc:
```diff
@@ -300,51 +300,74 @@ Using the append and prepend methods may invalidate an existing JSON document.
-append(docid, fragment)
-prepend(docid, fragment)
+append(document);
+prepend(document);
```
This needs a bit more, these samples don't make sense without the corresponding code to create the JsonDocument.

manishbansal8843 replied:
I feel the code samples included as part of this section will be enough to explain `append` method. After that, `prepend` method example would be exactly similar. Let me know your thoughts on it.

manishbansal8843 replied:
Also, compilation errors of the java snippets have been corrected.

programmatix commented on modules/ROOT/pages/core-operations.adoc:
```diff
@@ -300,51 +300,74 @@ Using the append and prepend methods may invalidate an existing JSON document.
+byteBuf toReplace = read1.content().writeBytes(toAppend);
+
+// Replace the existing document
+bucket.replace("binaryDoc",BinaryDocument.create("binaryDoc", toReplace));
```
I don't think this will compile.  Please do check that all examples compile and work.

manishbansal8843 replied:
Apart from small `b` in `byteBuff` in line 332, all lines got compiled successfully.

programmatix commented on modules/ROOT/pages/core-operations.adoc:
```diff
@@ -300,51 +300,74 @@ Using the append and prepend methods may invalidate an existing JSON document.
-# Append a fragment
-cb.append('binary_doc', '\x02', format=couchbase.FMT_BYTES)
+// Print it
+System.out.println(read.content().toString(CharsetUtil.UTF_8)); //Hello World
```
I doubt it'll be "Hello World" as there's no space in the appended data.  Please do check that all examples do what they say they'll do.

manishbansal8843 replied:
Apart from missing `;`, all lines got compiled successfully. Same is corrected now.
Further, i have removed space between `Hello` and `World`. I hope its fine.

programmatix commented on modules/ROOT/pages/core-operations.adoc:
```diff
@@ -42,22 +42,22 @@ Because JSON is a structured format, it can be subsequently searched and queried
+bucket.insert(document);
+bucket.replace(document);
+bucket.get(docid);
+bucket.remove(docid);
```
As elsewhere, these examples are too incomplete, they need to include creating the JsonDocument.

Consider what would be helpful to you as a developer.  E.g. the get example could show `JsonDocument result = bucket.get(docid);` instead.  Feel free to go a little above and beyond the Python snippets, which are a bit bare-bones.

manishbansal8843 replied:
Details explanations have already been given in relevant section on page https://docs.couchbase.com/java-sdk/2.7/document-operations.html. Will mention the links for reference.

programmatix commented on modules/ROOT/pages/core-operations.adoc:
```diff
@@ -42,22 +42,22 @@ Because JSON is a structured format, it can be subsequently searched and queried
-* _upsert_ will always replace the document, ignoring whether the ID has already existed or not.
+* _insert_ will only create the document if the given document with ID is not found within the database.
+* _replace_ will only replace the document if the given document with ID already exists within the database.
+* _upsert_ will always replace the document, ignoring whether the document with ID exists or not.
```
I'd word this "the given document's ID" instead.

manishbansal8843 replied:
yes. makes sense. Will correct.

manishbansal8843 replied:
Reason i mentioned `document with ID` because for insert/replace/update, we do not pass `documentId` as parameter. Instead we pass a `Document` object containing `documentId`

manishbansal8843 committed:
Incorporated few review comments.

manishbansal8843 committed:
Corrected java snippet for counter section

manishbansal8843 committed:
Corrected compilation errors 

Java snippets compilation errors are corrected for Raw byte concatenation section

manishbansal8843 reviewed (commented):
Kindly review again. All issues are resolved.

manishbansal8843 commented:
@programmatix Any updates on this PR. Does it need anymore edits?

osfameron commented:
Thanks @manishbansal8843 for this contribution for 2.7 docs, it looks like though it wasn't merged as is, it has informed the documentation of `array_add_unique`, `array_prepend` and friends in the 3.0 docs, see e.g. https://docs.couchbase.com/java-sdk/current/howtos/subdocument-operations.html#array-append-and-prepend

I'll close this PR now, thanks again!
